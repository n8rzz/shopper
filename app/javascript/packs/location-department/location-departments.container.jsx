import React from 'react';
import PropTypes from 'prop-types';
import _keyBy from 'lodash/keyBy';
import ApiService from '../service/api.service';
import { SortedDepartmentList } from './sorted-department-list.component';
import { NonSortedDepartmentList } from './non-sorted-department-list.component';

export class LocationDepartmentsContainer extends React.Component {
    constructor(props) {
        super(props);

        this._departmentMap = _keyBy(props.departments, 'id');
        this._onClickChangeDirectionHandler = this._onClickChangeDirection.bind(this);
        this._onClickCreateLocationDepartmentHanlder = this._onClickCreateLocationDepartment.bind(this);
        this._onClickRemoveHandler = this._onClickRemove.bind(this);
        this._onUpdateSortOrderSuccessHandler = this._onUpdateSortOrderSuccess.bind(this);
        this._onCreateLocationDepartmentHandler = this._onCreateLocationDepartment.bind(this);
        this._onDestroyLocationDepartmentHandler = this._onDestroyLocationDepartment.bind(this);

        this.state = {
            unsortedDepartmentList: this._buildUnsortedDepartmentsList(props.departments, props.locationDepartments),
            sortedDepartmentList: props.locationDepartments,
        };
    }

    componentWillUnmount() {
        this._onClickChangeDirectionHandler = null;
        this._onClickCreateLocationDepartmentHanlder = null;
        this._onClickRemoveHandler = null;
        this._onUpdateSortOrderSuccessHandler = null;
        this._onCreateLocationDepartmentHandler = null;
        this._onDestroyLocationDepartmentHandler = null;
    }

    _buildUnsortedDepartmentsList(departments, locationDepartments) {
        return departments.reduce((sum, department) => {
            const foundDepartment = locationDepartments.filter(
                (locationDepartment) => locationDepartment.department_id === department.id,
            );

            if (foundDepartment.length > 0) {
                return sum;
            }

            return [
                ...sum,
                department,
            ];
        }, []);
    }

    _onClickChangeDirection(locationDepartment, changeDirection) {
        this._updateSortOrder(locationDepartment, changeDirection);
    }

    _onClickCreateLocationDepartment(event, department) {
        event.preventDefault();

        const url = '/location_departments.json';
        const locationDepartmentParams = {
            location_id: this.props.locationId,
            department_id: department.id,
        };

        ApiService.post(url, locationDepartmentParams, this.props.csrf)
            .then((response) => this._onCreateLocationDepartmentHandler(response.data))
            .catch((error) => { throw error; });
    }

    // FIXME: combine this with identical method
    _onCreateLocationDepartment(locationDepartments) {
        const unsortedDepartmentList = this._buildUnsortedDepartmentsList(this.props.departments, locationDepartments);

        this.setState({
            unsortedDepartmentList,
            sortedDepartmentList: locationDepartments,
        });
    }

    _onClickRemove(locationDepartment) {
        const url = `/location_departments/${locationDepartment.id}.json`;

        ApiService.delete(url, null, this.props.csrf)
            .then((response) => this._onDestroyLocationDepartmentHandler(response.data))
            .catch((error) => { throw error; });
    }

    // FIXME: combine this with identical method
    _onDestroyLocationDepartment(locationDepartments) {
        const unsortedDepartmentList = this._buildUnsortedDepartmentsList(this.props.departments, locationDepartments);

        this.setState({
            unsortedDepartmentList,
            sortedDepartmentList: locationDepartments,
        });
    }

    _updateSortOrder(locationDepartment, changeDirection) {
        const url = `/location_departments/${locationDepartment.id}/${changeDirection.toLowerCase()}.json`;

        ApiService.post(url, null, this.props.csrf)
            .then((response) => this._onUpdateSortOrderSuccessHandler(response.data))
            .catch((error) => { throw error; });
    }

    _onUpdateSortOrderSuccess(locationDepartments) {
        this.setState({ sortedDepartmentList: locationDepartments });
    }

    render() {
        return (
            <React.Fragment>
                <div className={'vr_3'}>
                    <SortedDepartmentList
                        departmentMap={this._departmentMap}
                        locationDepartments={this.state.sortedDepartmentList}
                        onClickChangeDirectionHandler={this._onClickChangeDirectionHandler}
                        onClickRemoveHandler={this._onClickRemoveHandler}
                    />
                </div>
                <div className={'vr_2'}>
                    <NonSortedDepartmentList
                        departments={this.state.unsortedDepartmentList}
                        onClickCreateLocationDepartmentHanlder={this._onClickCreateLocationDepartmentHanlder}
                    />
                </div>
            </React.Fragment>
        );
    }
}

LocationDepartmentsContainer.propTypes = {
    csrf: PropTypes.string.isRequired,
    departments: PropTypes.array.isRequired,
    locationDepartments: PropTypes.array.isRequired,
    locationId: PropTypes.number.isRequired,
};
