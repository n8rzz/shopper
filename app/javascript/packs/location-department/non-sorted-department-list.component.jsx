import React from 'react';
import PropTypes from 'prop-types';

export function NonSortedDepartmentList(props) {
    const departmentListJsx = props.departments.map((department, index) => (
        <li
            className={'listItem mix-listItem_tight'}
            key={`nonSorted-department-${index}`}
        >
            <div className={'listItem-hd'}>
                <ul className={'stereo'}>
                    <li>
                        <a
                            className={'link'}
                            href={`/departments/${department.id}`}
                        >
                            {department.name}
                        </a>
                    </li>
                    <li>
                        <button
                            className={`btn js-addDepartmentToSort-${department.id}`}
                            type={'button'}
                            onClick={(event) => props.onClickCreateLocationDepartmentHanlder(event, department)}
                        >
                            ADD
                        </button>
                    </li>
                </ul>
            </div>
        </li>
    ));

    return (
        <React.Fragment>
            <div className={'vr_1'}>
                <h3 className={'hdg hdg_3'}>Unsorted Departments</h3>
            </div>
            <ul className={'vlist js-nonSortedDepartmentList'}>
                { departmentListJsx }
            </ul>
        </React.Fragment>
    );
}

NonSortedDepartmentList.propTypes = {
    departments: PropTypes.array.isRequired,
    onClickCreateLocationDepartmentHanlder: PropTypes.func.isRequired,
};
