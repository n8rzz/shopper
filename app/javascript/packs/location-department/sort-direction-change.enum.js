export const SORT_DIRECTION_CHANGE = {
    DOWN: 'DOWN',
    UP: 'UP',
};
