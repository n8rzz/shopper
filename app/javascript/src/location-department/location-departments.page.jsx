import React from 'react';
import PropTypes from 'prop-types';
import { LocationDepartmentsContainer } from './location-departments.container';

export function LocationDepartmentsPage(props) {
    return (
        <LocationDepartmentsContainer
            csrf={props.csrf}
            departments={props.departments}
            locationDepartments={props.locationDepartments}
            locationId={props.locationId}
        />
    );
}

LocationDepartmentsPage.propTypes = {
    csrf: PropTypes.string.isRequired,
    departments: PropTypes.array.isRequired,
    locationDepartments: PropTypes.array.isRequired,
    locationId: PropTypes.number.isRequired,
};
