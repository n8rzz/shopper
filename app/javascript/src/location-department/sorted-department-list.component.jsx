import React from 'react';
import PropTypes from 'prop-types';
import { SORT_DIRECTION_CHANGE } from './sort-direction-change.enum';

export class SortedDepartmentList extends React.Component {
    constructor(props) {
        super(props);

        this._onClickUpHandler = this._onClickUp.bind(this);
        this._onClickDownHandler = this._onClickDown.bind(this);
        this._onClickRemoveHandler = this._onClickRemove.bind(this);
    }

    componentWillUnmount() {
        this._onClickUpHandler = null;
        this._onClickDownHandler = null;
        this._onClickRemoveHandler = null;
    }

    _onClickUp(event, locationDepartment) {
        event.preventDefault();

        this.props.onClickChangeDirectionHandler(locationDepartment, SORT_DIRECTION_CHANGE.UP);
    }

    _onClickDown(event, locationDepartment) {
        event.preventDefault();

        this.props.onClickChangeDirectionHandler(locationDepartment, SORT_DIRECTION_CHANGE.DOWN);
    }

    _onClickRemove(event, locationDepartment) {
        event.preventDefault();

        this.props.onClickRemoveHandler(locationDepartment);
    }

    render() {
        const departmentsCount = this.props.locationDepartments.length - 1;
        const departmentListJsx = this.props.locationDepartments.map((locationDepartment, index) => {
            const isUpButtonDisabled = locationDepartment.sort_order === 0;
            const isDownButtonDisabled = index === departmentsCount;
            const department = this.props.departmentMap[locationDepartment.department_id];

            return (
                <li
                    className={'listItem'}
                    key={`sorted-department-${index}`}
                >
                    <div className={'listItem-hd'}>
                        <ul className={'stereo'}>
                            <li>
                                <a
                                    className={'link'}
                                    href={`/departments/${department.id}`}
                                >
                                    {department.name}
                                </a>
                            </li>
                            <li>
                                <button
                                    className={`icon-btn js-sortedDepartment-remove-${locationDepartment.id}`}
                                    type={'button'}
                                    onClick={(event) => this._onClickRemoveHandler(event, locationDepartment)}
                                >
                                    <span className={'icon icon-trashcan'} />
                                </button>
                            </li>
                        </ul>
                    </div>
                    <div className={'listItem-ft'}>
                        <ul className={'hlist'}>
                            <li>
                                <button
                                    className={`btn js-sortedDepartment-up-${locationDepartment.id}`}
                                    type={'button'}
                                    disabled={isUpButtonDisabled}
                                    onClick={(event) => this._onClickUpHandler(event, locationDepartment)}
                                >
                                    UP
                                </button>
                            </li>
                            <li>
                                <button
                                    className={`btn js-sortedDepartment-down-${locationDepartment.id}`}
                                    type={'button'}
                                    disabled={isDownButtonDisabled}
                                    onClick={(event) => this._onClickDownHandler(event, locationDepartment)}
                                >
                                    DOWN
                                </button>
                            </li>
                        </ul>
                    </div>
                </li>
            );
        });

        return (
            <React.Fragment>
                <div className={'vr_1'}>
                    <h3 className={'hdg hdg_3'}>Sort Order</h3>
                </div>
                <ul className={'vlist js-sortedDepartmentList'}>
                    { departmentListJsx }
                </ul>
            </React.Fragment>
        );
    }
}

SortedDepartmentList.propTypes = {
    departmentMap: PropTypes.object.isRequired,
    locationDepartments: PropTypes.array.isRequired,
    onClickChangeDirectionHandler: PropTypes.func.isRequired,
    onClickRemoveHandler: PropTypes.func.isRequired,
};
