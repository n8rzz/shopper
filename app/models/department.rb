# frozen_string_literal: true

# id:           integer
# name:         string
# created_at:   datetime
# updated_at:   datetime
class Department < ApplicationRecord
  belongs_to :ownable, polymorphic: true

  has_many :items, dependent: :nullify
  has_many :order_items, dependent: :nullify
  has_many :orders, through: :order_items, dependent: :nullify
  has_many :location_departments, dependent: :nullify
  has_many :locations, through: :location_departments, dependent: :nullify

  validates :name, presence: true, uniqueness: true

  scope :by_name, -> { order('name ASC') }
end
