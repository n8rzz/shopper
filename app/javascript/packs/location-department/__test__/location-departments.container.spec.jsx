import React from 'react';
import { shallow } from 'enzyme';
import { csrfMock } from '../../__mocks__/csrf.mock';
import { departmentListMock } from '../../__mocks__/department.mock';
import { locationDepartmentListMock } from '../../__mocks__/location-department.mock';
import { LocationDepartmentsContainer } from '../location-departments.container';

describe('LocationDepartmentsContainer', () => {
    let component = null;

    beforeEach(() => {
        component = shallow(<LocationDepartmentsContainer
            csrf={csrfMock}
            departments={departmentListMock}
            locationDepartments={locationDepartmentListMock}
            locationId={1}
        />);
    });

    afterEach(() => {
        component = null;
    });

    test('renders correctly with valid props', () => {
        expect(component).toMatchSnapshot();
    });
});
