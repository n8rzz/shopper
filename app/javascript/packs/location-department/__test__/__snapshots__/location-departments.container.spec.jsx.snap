// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LocationDepartmentsContainer renders correctly with valid props 1`] = `
<Fragment>
  <div
    className="vr_3"
  >
    <SortedDepartmentList
      departmentMap={
        Object {
          "3": Object {
            "created_at": "2019-07-21T16:03:32.883Z",
            "id": 3,
            "name": "Bakery",
            "updated_at": "2019-07-21T16:03:32.883Z",
          },
          "5": Object {
            "created_at": "2019-07-21T16:03:32.883Z",
            "id": 5,
            "name": "Meat",
            "updated_at": "2019-07-21T16:03:32.883Z",
          },
        }
      }
      locationDepartments={
        Array [
          Object {
            "created_at": "2019-08-22T00:58:38.793Z",
            "department_id": 3,
            "id": 1,
            "location_id": 1,
            "sort_order": 0,
            "updated_at": "2019-08-22T00:58:38.793Z",
          },
          Object {
            "created_at": "2019-08-22T00:58:38.793Z",
            "department_id": 5,
            "id": 2,
            "location_id": 1,
            "sort_order": 0,
            "updated_at": "2019-08-22T00:58:38.793Z",
          },
        ]
      }
      onClickChangeDirectionHandler={[Function]}
      onClickRemoveHandler={[Function]}
    />
  </div>
  <div
    className="vr_2"
  >
    <NonSortedDepartmentList
      departments={Array []}
      onClickCreateLocationDepartmentHanlder={[Function]}
    />
  </div>
</Fragment>
`;
